src/xtgeo/xyz/_xyz_io.py

```diff
@@ -11,140 +14,372 @@
 
 logger = xtg.functionlogger(__name__)
 
+
+class _ValidDataFrame(pd.DataFrame):
+    ...
+
+
 # -------------------------------------------------------------------------
-# Import/Export methods for various formats
-# Note: 'self' is the XYZ instance which may be Points/Polygons
+# Import/Export methods for various formats; primarely class methods
 # -------------------------------------------------------------------------
 
 
-def import_xyz(self, pfile, zname="Z_TVDSS"):
-    """Simple X Y Z file. All points as Pandas framework."""
+def import_xyz(pfile, zname="Z_TVDSS", is_polygons=False):
+    """Simple text X Y Z files on rms_ascii/irap_text or just XYZ.
 
-    self.zname = zname
+    The format names 'irap_ascii', 'irap_text', 'rms_ascii', 'rms_text' are all the
+    same format (the name has changed through history; the original was "irap_ascii"),
+    while XYZ differs somewhat.
 
-    self._df = pd.read_csv(
-        pfile,
-        delim_whitespace=True,
-        skiprows=0,
-        header=None,
-        names=[self._xname, self._yname, zname],
-        dtype=np.float64,
-        na_values=999.00,
-    )
+    The are some subtle differences between a RMS text file and XYZ file
+    according to RMS documentation:
+
+    Points:
+
+    rms_ascii and XYZ seems identical!
+
+    Polygons:
+
+    rms_ascii have 3 columns and will contain a line 999.0 999.0 999.0 to
+    seperate polygons.
+
+    The XYZ files valid for Polygons will have an ID column. It may be 3 or 4 columns,
+    if only 3 then the Z values is assumed to be 0.0. The last column is always the ID.
+    """
+    assume_rms_text = True  # assume that XYZ file is irap text based
+
+    ncolumns = 3
+    with open(pfile.file, "r", encoding="utf8") as stream:
+        fields = stream.readline().split()
+        if len(fields) == 4:
+            assume_rms_text = False
+            ncolumns = 4
+            if not is_polygons:
+                raise TypeError(
+                    "Points is assumed but this appears to be a XYZP "
+                    "for Polygons as the file has 4 columns"
+                )
+        else:
+            if not is_polygons:
+                assume_rms_text = True
+            else:
+                assume_rms_text = False
+
+            if "." in fields[2]:
+                assume_rms_text = True
+
+    kwargs = {}
+    _xn = kwargs["xname"] = "X_UTME"
+    _yn = kwargs["yname"] = "Y_UTMN"
+    _zn = kwargs["zname"] = zname
+
+    if assume_rms_text:
+        dfr = pd.read_csv(
+            pfile.file,
+            delim_whitespace=True,
+            skiprows=0,
+            header=None,
+            names=[_xn, _yn, _zn],
+            dtype=np.float64,
+            na_values=999.00,
+        )
+    if not is_polygons:
+        dfr.dropna(inplace=True)
```

Comment:
This would fail if `not assume_rms_text and not is_polygons`.

Reply:
Missed an indent :+1: